File: examples/cpp-examples/MeshDecimate.dox.cpp
#include <MRMesh/MRMeshFwd.h>
#include <MRMesh/MRMeshLoad.h>
#include <MRMesh/MRMeshSave.h>
#include <MRMesh/MRMesh.h>
#include <MRMesh/MRMeshDecimate.h>
#include <MRMesh/MRBuffer.h>

#include <iostream>

int main()
{
    // Load mesh
    auto meshRes = MR::MeshLoad::fromAnySupportedFormat( "mesh.stl" );
    if ( !meshRes.has_value() )
    {
        std::cerr << meshRes.error() << std::endl;
        return 1;
    }

    MR::Mesh& mesh = *meshRes;
    
//! [0]
    // Repack mesh optimally.
    // It's not necessary but highly recommended to achieve the best performance in parallel processing
    mesh.packOptimally();

    // Setup decimate parameters
    MR::DecimateSettings settings;

    // Decimation stop thresholds, you may specify one or both
    settings.maxDeletedFaces = 1000; // Number of faces to be deleted
    settings.maxError = 0.05f; // Maximum error when decimation stops

    // Number of parts to simultaneous processing, greatly improves performance by cost of minor quality loss.
    // Recommended to set to the number of available CPU cores or more for the best performance
    settings.subdivideParts = 64;

    // Decimate mesh
    MR::decimateMesh( mesh, settings );
//! [0]

    // Save result
    if ( auto saveRes = MR::MeshSave::toAnySupportedFormat( mesh, "decimated_mesh.stl" ); !saveRes )
    {
        std::cerr << saveRes.error() << std::endl;
        return 1;
    }

    return 0;
}

File: examples/python-examples/MeshDecimate.dox.py
import meshlib.mrmeshpy as mrmeshpy

# Load mesh
mesh = mrmeshpy.loadMesh("mesh.stl")

# Repack mesh optimally.
# It's not necessary but highly recommended to achieve the best performance in parallel processing
mesh.packOptimally()

# Setup decimate parameters
settings = mrmeshpy.DecimateSettings()

# Decimation stop thresholds, you may specify one or both
settings.maxDeletedFaces = 1000 # Number of faces to be deleted
settings.maxError = 0.05 # Maximum error when decimation stops

# Number of parts to simultaneous processing, greatly improves performance by cost of minor quality loss.
# Recommended to set to number of CPU cores or more available for the best performance
settings.subdivideParts = 64

# Decimate mesh
mrmeshpy.decimateMesh(mesh, settings)

# Save result
mrmeshpy.saveMesh(mesh, "decimatedMesh.stl")
